From 5be94cf28f112a2f570e50acdc738de1796ca00f Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert@linux-m68k.org>
Date: Tue, 18 Nov 2008 21:25:15 +0100
Subject: [PATCH 34/90] m68k: Fix debug=mem on Amiga

`debug=mem' on Amiga has been broken for a while.
early_param() processing is done very/too early, i.e. before
amiga_hw_present.CHIP_RAM has been set in amiga_identify(), causing
amiga_savekmsg_setup() not to find any Chip RAM.

The quick hack below makes it work again by introducing an additional
intermediate flag, but I don't like it.

Signed-off-by: Geert Uytterhoeven <geert@linux-m68k.org>

TODO: Suggestion from Roman Zippel <zippel@linux-m68k.org>:

The alternative to pull some of the setup ahead of the early_param() call.
It shouldn't be a big problem to move the call after the config calls, as
these should only do a very basic setup, everything else should be done
via init calls (e.g. amiga_init_sound() might be such a candidate).
---
 arch/m68k/amiga/config.c |   30 +++++++++++++++++++++---------
 1 files changed, 21 insertions(+), 9 deletions(-)

diff --git a/arch/m68k/amiga/config.c b/arch/m68k/amiga/config.c
index 6e56275..71232cd 100644
--- a/arch/m68k/amiga/config.c
+++ b/arch/m68k/amiga/config.c
@@ -106,6 +106,7 @@ static void amiga_reset(void);
 extern void amiga_init_sound(void);
 static void amiga_mem_console_write(struct console *co, const char *b,
 				    unsigned int count);
+static void __init amiga_savekmsg_init(void);
 #ifdef CONFIG_HEARTBEAT
 static void amiga_heartbeat(int on);
 #endif
@@ -116,6 +117,8 @@ static struct console amiga_console_driver = {
 	.index	= -1,
 };
 
+static int __initdata amiga_enable_debug_mem;
+
 
     /*
      *  Motherboard Resources present in all Amiga models
@@ -461,6 +464,9 @@ void __init config_amiga(void)
 	/* initialize chipram allocator */
 	amiga_chip_init();
 
+	if (amiga_enable_debug_mem && AMIGAHW_PRESENT(CHIP_RAM))
+		amiga_savekmsg_init();
+
 	/* our beloved beeper */
 	if (AMIGAHW_PRESENT(AMI_AUDIO))
 		amiga_init_sound();
@@ -780,18 +786,10 @@ static void amiga_mem_console_write(struct console *co, const char *s,
 	}
 }
 
-static int __init amiga_savekmsg_setup(char *arg)
+static void __init amiga_savekmsg_init(void)
 {
 	static struct resource debug_res = { .name = "Debug" };
 
-	if (!MACH_IS_AMIGA || strcmp(arg, "mem"))
-		goto done;
-
-	if (!AMIGAHW_PRESENT(CHIP_RAM)) {
-		printk("Warning: no chipram present for debugging\n");
-		goto done;
-	}
-
 	savekmsg = amiga_chip_alloc_res(SAVEKMSG_MAXMEM, &debug_res);
 	savekmsg->magic1 = SAVEKMSG_MAGIC1;
 	savekmsg->magic2 = SAVEKMSG_MAGIC2;
@@ -800,6 +798,20 @@ static int __init amiga_savekmsg_setup(char *arg)
 
 	amiga_console_driver.write = amiga_mem_console_write;
 	register_console(&amiga_console_driver);
+}
+
+static int __init amiga_savekmsg_setup(char *arg)
+{
+	if (!MACH_IS_AMIGA || strcmp(arg, "mem"))
+		goto done;
+
+	if (!AMIGAHW_PRESENT(CHIP_RAM)) {
+		printk("Warning: no chipram present for debugging\n");
+		amiga_enable_debug_mem = 1;
+		goto done;
+	}
+
+	amiga_savekmsg_init();
 
 done:
 	return 0;
-- 
1.6.2.3

